This collection of examples serves as a valuable reference for utilizing the Jupyter Notebook in the context of mechanical design. It not only demonstrates how to perform calculations, but also delves into the crucial aspect of conducting research in this field. The code closely aligns with the content found in the Roloff & Matek textbook's 6th edition, referencing equation and expression numbers.  

We strongly encourage you to not only test but also customize this code to align with your unique requirements and approach. To enhance your understanding, we've established a straightforward differentiation between comments dedicated to grasping mechanical design principles and those designed to assist you in navigating the code. You'll easily spot the latter category as they will be presented in italic font. Additionally, within coding blocks, comments relating to code usage will be clearly marked as follows:  

```
#--- Comments on how to use the code will have 3 minus signs to the left and right ---
```
These formatting conventions will help streamline your experience while working with the code. 

To furhter enhance your experience, it's essential to master the art of creating Markdown cells effectively. You can refer [website on the basicis of Markdown](https://www.markdownguide.org/basic-syntax/) basics to discover how to craft visually appealing Markdown text within your code
You will also have to learn a proper way to create markdown cells.  For a more detailed explenation on the syntax on using Markdown, you can visit the [jupyter notebook website](https://jupyter-notebook.readthedocs.io/en/latest/examples/Notebook/Working%20With%20Markdown%20Cells.html) .

---

# Design of chain drive: 'Top-Down, equations first apporach'

*There are two primary ways to utilize the code effectivelyµ. The first approach begins by identifying the final equation of interest and generating a symbolic expression. Gradually, you can substitute the various symbols with actual values. Several methods will help you to keep track on the parameters to be filled in.*  
*We will refer to this as the "top-down" approach.*

*The second approach, which is detailed in another example file, resembles using a pocket calculator, where you follow a sequential order to compute expressions, ultimately culminating in the final result. In this method, all the parameters within the calculation chain maintain numerical values throughout.*  
*For a more detailed explanation, please refer to the other example, which will provide further elaboration.*
*We will refer to this as the "bottom-up" approach.*  

*Tip: You can easily enable or disable chapter header numbering by accessing the table of contents (View -> Table of Contents) and clicking on the small icon located just below the document's name in the table of contents.*

---
**Within this document, we will employ a top-down, equation-first methodology. This approach commences by focusing on the primary expression and employing it as a guide to pinpoint missing parameters. Subsequently, the next phase involves populating these parameters. In the concluding step, we explicitly substitute these parameters into the primary expression.**  
**Notably, this process can be recursive if one of the parameters necessitates the evaluation of another expression.**  


**While this document predominantly adheres to the top-down approach as an illustrative example, feel free to explore the possibility of blending both bottom-up and top-down methods to suit your specific needs.**  
**We strongly encourage you to adapt the methodology to your own thought process and preferences.**  

---

# Assignment

In this assignment, we will design the drive system for a transport belt. The drive will be powered by a geared electromotor and a chain transmission. Here are the initial specifications:

- The output power of the geared motor is given as $P_1 = 3 \: kW$ at an output speed of $n_1 = 125 \:rpm$.
- The desired speed of the shaft driving the transport belt is approximately $n_2 \approx 50 \:rpm$.
- The distance between the output shaft of the geared motor and the shaft driving the transport belt should be approximately $a_0 \approx 1000 \: mm$, with an angle between the shaft-to-shaft line and the horizontal of $\delta \approx 40°$
- For this specific application of the transport belt, we will assume the following use conditions: average starting, full load with moderate jolts, 8 hours of daily use, and a well-lubricated chain.

Your task is to calculate the following:

  a. Determine the number of teeth, $z_1$ and $z_2$, required for the chainwheels.  
  b. Select a suitable roller chain according to DIN 8187.  
  c. Identify the appropriate lubrication to be used.  
  d. Calculate the forces $F_A$ acting on the drive shaft.  

---
 

In [1]:
# import statements
import sympy as sp
import numpy as np
import matplotlib.pyplot as plt

import MechDesign.Helpers as HM

from MechDesign.Units.Units import m_, mm_, kg_, s_, N_, rpm_, W_, deg_
import MechDesign.Units.UnitMethods as UM

import MechDesign.RnM as RnM

*In this section, a set of import statements is being made, and this block will be almost identical for all of your files. The only line to change is the one defining the units that you will be using throughout the rest of the document.*  

*You are advised to start off with only the units that you are sure you will be needing, e.g., m_*

*By definition, all units have a trailing '_'. Running the following code will return a list of units that are defined. If you need additional units, please contact your teaching staff.*  

```
UM.PrintAvailableUnits()
```

*Aside from the units themselves, several helper functions are also defined. These functions can be used to remove or convert units. Running the following line of code will return a list of available helper functions.*

```
help(UM)
```


In [2]:
# --- comment out the '#' at the beginning of the lines to run the code mentioned above ---
#UM.PrintAvailableUnits()
#help(UM)
#help(RnM)         #--- calling this line will give you an overview of the chapters available in this package ---

## a. Gear Ratio
By utilizing the required speeds, we can calculate the gear ratio, denoted as $i$. In a first step we will estimate $i'$.  The final gear ratio will depend on the number of teeth of both chain wheels. 

*To calculate the gear ratio, the initial step is to create a 'new chain calculation.' This process automatically assigns a single symbol to each of the necessary parameters outlined in the Chain Calculation chapter. Once these symbols are defined, we can proceed to input the known values, effectively replacing the symbols with their respective values.*  
*In the code section below, you will also find some code to demonstrate that this newly created object indeed contains all of these symbols.*


In [3]:
# creating the chain to be calculated
CH = RnM.Chain()     #--- since Chain is part of the RnM library we call it as a part of it ---

#--- code to demonstrate that CH holds the parameters ---
print(CH.i)          #---This prints the raw value of a symbol named '_i,' a form you are unlikely to use ---
display(CH.i)        #---This prints (displays) a better readable form of symbol '_i' ---
print(CH.n_1)        #---This prints the raw value of a symbol named '_n_1,' a form you are unlikely to use ---
display(CH.n_1)      #---A better readable form of symbol '_n_1', note how the use of the '_' before the '1' generates a nice subscript ---

#---When you're inside a code block (e.g. on the line), try typing 'CH.' and then press 'TAB.' This will display a list of the created parameters.---

_i


_i

_n_1


_n_1

In [4]:
#---In the context of this mechanical design code, a distinct class of symbols has been defined. ---
#---These symbols can include 'a description,' 'a comment,' 'a default value,' and 'a range.' ---
#---Note that the key distinction between 'description' and 'comment' lies in their usage: the former is fixed, ----
#---while the latter can be freely utilized for making notes. Let's demonstrate---
#---When attempting this on your own, don't forget to take advantage of tab-completion. ---
#---Note that tab-completion only functions once the code has been executed. ----
display(CH.i.description)
display(CH.n_1.description)
CH.i.comment = 'we still need to figure out this value'
CH.i.range = [1, 5000]
CH.i.def_value = 5678

'[] gear ratio'

'[] speed of driving chain wheel'

In [5]:
#---Several helper functions are also available, including one for printing out expressions. All of these functions are part of the 'H' module. ---
HM.MyHelp(CH.i)       #---The function is named 'MyHelp' because it provides assistance for the expressions you generate.---
#---Let's create a very simple, nonsensical dummy expression ---
DummyExpression = CH.i*CH.n_1
HM.MyHelp(DummyExpression)
#---Note that symbols are indeed closley printed together.---
#---Utilize these additional properties to your advantage while coding.----

_i

    i : [] gear ratio
          we still need to figure out this value
          default value: 5678
          range: 1;5000


_i*_n_1

    n_1 : [] speed of driving chain wheel
    i : [] gear ratio
          we still need to figure out this value
          default value: 5678
          range: 1;5000


In [6]:
#---determining the main expression: not a lot of calculation is needed and we could identify the final gear ration as the main expression ---
#---Probably, this is good of example where a mixed Top-Down & Bottum-Up would work better ---
#---In this particular scenario, we have the option to manually code this, but all the formulas from the book are already available.---
CH.i = CH.E17_1B_GearRatioTeeth() #---When multiple expressions share the same equation number, they are ordered from left to right as A,B,C, etc.---
#---Again remember the tab-completions, expressions all start with E'chapterNr'_'NrOfExpression. ---
#---Typing 'E17_1' and pressing 'Tab' will limit the options, and tab-completion will help you avoid typos in the method names ---
#---Let's look at what we have now ---
HM.EqPrint('i',CH.i)     #---The first input of 'EqPrint' can be any text without spaces and will be placed on the left-hand side of the equation.---
HM.EqPrint('something',CH.i)
t=0 #---By default, code cells will print the last line of code if it doesn't contain an '=' sign. Adding 't=0' is just to prevent double printing.---

Eq(i, _z_2/_z_1)

Eq(something, _z_2/_z_1)

In [7]:
#---An alternative approach for examining the expression is to utilize the HM.MyHelp method. ---
HM.MyHelp(CH.i)
t=0
#---As you can observe, this already offers valuable insights into the parameters that require input. ---
#---In cases where parameters are sourced from a table, you will also find clear indications within the comments. ---

_z_2/_z_1

    z_2 : [] nr of teeth of the driven chain wheel
    z_1 : [] nr of teeth of the driving chain wheel


we have to determine the values of $z_2$ and $z_1$.  Based on the textbook, we need an intial estimation of the required gear ratio, which we will call $i'$

In [8]:
# setting the value of known parameters
CH.n_1=125 * rpm_    #---Notice the imidiate addition of units---
CH.n_2=50 * rpm_
iprime = CH.n_1/CH.n_2
t=HM.EqPrint('iprime',iprime)
#---Notice that the two units, of course, cancel each other out. ---

Eq(iprime, 2.5)

For this transmission, the number of teeth mentioned in section 17.2.2 come closest to standard roller chain sizes: $z_1=23$ for the small wheel (selected from the range $z=17...25$, which is favorable for small wheels), and $z_2=57$ for the large wheel (determined by multiplying $z_1=23$ by $i'$ ).  
Therefore, the actual gear ratio becomes $i = z_2/z_1$

In [9]:
CH.z_1 = 23
CH.z_2 = 57
#---Recall that we've previously established the relationship: CH.i = z_2/z_1 in one of the code blocks above. ---
#---Setting the expression first, and then "Substituting variables" is the crux of the top-down approach. ---
#---To facilitate the substitution of variables, we need to specify the exact instance of the CHAIN type from which we extract these variables. ---
#---This could be usefull if you are calculating multiple chains, and have one set of expressions. Comparing two options could be a usecase ---
#---For instructional purposes, we will initially store the outcome in a temporary variable ---
#---The first element to be inserted into the 'substitute' method is the symbolic expression, the second is the instance of the chain to be used ---
#---Take note that we employ 'CH' as the second input: it refers to the entire chain, encompassing all its variables. ---
i_result = HM.substitute(CH.i,CH)   
HM.EqPrint('i_result',i_result)
#---By storing the result in a separate variable, we can observe that the original expression remains unaltered. ---
t=HM.EqPrint('i',CH.i)

Eq(i_result, 2.478)

Eq(i, _z_2/_z_1)

In [10]:
#---Let's demonstrate a more advanced way of using the 'substitute' method. ---
#---We can overrule one or more of the CH variabels, by explicitly stating what to replace the value with. ---
#---Suppose we want to investigate the effect of a different value for z_2.---
i_result2 = HM.substitute(CH.i,CH,z_2=114)   #---CH.z_2 is not changed, the "alternative" z_2 is only used for this spefic calculation. ---
HM.EqPrint('i_result2',i_result2)
HM.EqPrint('z_2',CH.z_2)  #--- as you can see CH.z_2 remains unchanged. ---
#---Or replacing both values and without an additional line of code. ---
t=HM.EqPrint('i_result3',HM.substitute(CH.i,CH,z_2=114,z_1=10))   #---When using this "one line of code" the result is only printed, not stored. ---

Eq(i_result2, 4.957)

Eq(z_2, 57)

Eq(i_result3, 11.4)

In [11]:
#---A more typical scenario would involve replacing the expression within CH.i with its updated value. ---
#---First, let's demonstrate again that CH.i is still an expression. ---
HM.MyHelp(CH.i)
#---Replacing the expression with specific values, based on CH. ---
CH.i = HM.substitute(CH.i,CH)
t=HM.EqPrint('i',CH.i)
#---Note that running this code block twice will cause the 'MyHelp' method to display the current value of 'i' instead of the symbolic expression. 
#---The symbolic expression of 'i' has indeed been fully replaced by the new value. ---
#---To regain access to the symbolic expression, you will need to rerun the entire notebook. ---

_z_2/_z_1

    z_2 : [] nr of teeth of the driven chain wheel
    z_1 : [] nr of teeth of the driving chain wheel


Eq(i, 2.478)

**The solutions to quesion a. is:**
- $z_1 = 23$ with this smaler chain wheel beeing the driving wheel
- $z_2 = 57$ with this bigger geer beeing the driven wheel
  
the gear ratio then becomes $i \approx  2.48$   
*Please be aware that the '2.48' value is hard-coded in this Markdown cell, and there is no straightforward method to incorporate values from the calculations directly within the Markdown content.*

---

## b. Chain selection

The selection of the chain size is made using the power diagram (Table 17-3). 
However, since the operating conditions here deviate from the values in the diagram, 
the desired diagram power must be determined according to equation 17.7.

In [12]:
# the main expression to evaluate is the calculation of P_D
CH.P_D = CH.E17_7_DiagramPower()
t=HM.MyHelp(CH.P_D)

_K_A*_P_1*_f_1/(_f_2*_f_3*_f_4*_f_5*_f_6)

    f_2 : [] power factor on shaft to shaft distance {table 17-6}
    f_3 : [] power factor on shackle shape {exp 17.7}
    P_1 : [W] input power
    f_1 : [] power factor on number of teeth {table 17-5}
    K_A : [] working condtion factor {table 3-5}
    f_5 : [] power factor on lifetime expectancy 
    f_4 : [] power factor on number of chain wheels {exp 17.7}
    f_6 : [] power factor on working conditions {table 17-7}


In [13]:
CH.K_A = 1.6     #The operating factor K_A , corresponding to the occurring conditions and according to Table 3-5b
CH.f_1 = 0.8     #Estimating the number of teeth, from Table 17-5, a factor of f_1 = 0.8 is read for z_1=23.
CH.f_2 = 1       #The shaft distance factor f_2 is initially set to 1 because the chain pitch is still unknown.
CH.f_3 = 1       #as we assume an even number of connecting links.
CH.f_4 = 1       #for n = 2
CH.f_5 = 1       #for an, assumed service life of 15000hours
CH.f_6 = 0.9     #according to table 17.7 a dust free environment cannot be guaranteed
#---Although it can be found in the assignment itself, we will keep 'P_1' as an unknown variable for instructional purposes. --
#CH.P_1 = 3000*W_ #based on the data in the assignment itself
#---Substitute the known variables into the expression and displaying what is still lacking. 
CH.P_D = HM.substitute(CH.P_D,CH)
t=HM.MyHelp(CH.P_D)

1.422*_P_1

    P_1 : [W] input power


*We now have a symbolic expression of $P_D=f(P_1)$, not in general but for this specific chain called 'CH'. Having such specific expression well comin handy when taking the next step towards plotting the impact of a paramater.*

In [14]:
CH.P_1 = 3000*W_ #based on the data in the assignment itself
CH.P_D = HM.substitute(CH.P_D,CH)
t = HM.EqPrint('P_D',CH.P_D)

Eq(P_D, 4267.0*W_)

For this power and for $n_1\approx125$ rpm, we select from Table 17-3:  
DIN 8187 - 16B - 1 : roller chain no. 16B, single strand. The pitch for this chain is $p=25.4$ mm.  
  
For the expected shaft distance $a_0\approx1000$ mm, the number of chain links can be calculated using equation 17.9.

In [15]:
CH.X_0 = CH.E17_9_InitalNrOfShackels()
CH.p = 25.4*mm_       #as foud for the DIN 8187-16B chain
CH.a_0 = 1000*mm_     #as per assignment
CH.X_0 = HM.substitute(CH.X_0,CH)
t=HM.EqPrint('X_0',CH.X_0)

Eq(X_0, 119.5)

Following the recommendations in paragraph 17.2.5, it is preferred to have an even number of links. Therefore, the number of links is set to $X=120$.  
The effective shaft distance, $a$, can be calculated using expression 17.10

In [16]:
#---As an illustrative example, we will combine both the Top-Down and Bottom-Up approaches. ---
#---The following code employs the Bottom-Up approach, initially defining the parameters and then establishing the expression. ---
CH.X = 120
CH.a = CH.E17_10_ShaftDistance()
HM.EqPrint('a',CH.a)              #---This output will still contain the symbol for 'pi' ---
t=HM.EqPrint('a',CH.a.evalf())    #---By adding 'ealf()' into the equation, we compel the expression to undergo a comprehensive evaluation.---

Eq(a, 6.35*mm_*(80.0*(1 - 0.36125/pi**2)**0.5 + 80.0))

Eq(a, 1007.0*mm_)

**The solution to question b. is:**

- DIN 8187 - 16B - 1x120, chain
- single chain
- 120 links
- shaft distance $a=1007$ mm  


---

## c. lubrication 

The choice of lubrication depends on both the speed $v$ and the chain size (refer to Table 17-8).  
We can determine the speed by considering the diameter $d$ of one of the sprockets, utilizing expression 17.3

In [17]:
# opting to calcuate the speed using the smallest sprocket, opting for expression 17.3B, thus eliminating the need to calculate tau
CH.v = CH.E17_14_h_Speed()  #---Helper expressions within or adjacent to an expression are indicated using '_h_' or '_hA_','_hB_','_hC_,..
HM.EqPrint('v',CH.v)
#since 3B can be used for both d_1 and d_2, z has to be set to z_1 or z_2
CH.d_1 = HM.substitute(CH.E17_3B_PitchDiameter(),CH,z=CH.z_1)       #---increasing code complexity by adding multiple steps in one line ---
HM.EqPrint('d_1',CH.d_1)
CH.v = HM.substitute(CH.v,CH)    #---It would probably make more sense to also use the Bottum-Up and put d_1 calculation before the CH.E17_14_h_speed()
t=HM.EqPrint('v',CH.v)

Eq(v, 62.5*_d_1*rpm_)

Eq(d_1, 186.5*mm_)

Eq(v, 11660.0*mm_*rpm_)

In [18]:
#---Use the 'UM.All_to_SI' method to perform a comprehensive unit conversion to the International System of Units (SI)---
CH.v = UM.All_to_SI(CH.v)
t=HM.EqPrint('v',CH.v)

Eq(v, 1.221*m_/s_)

Table 17-8 suggests that at a velocity of $v=1.22 m/s$ and with a size of 16B, drip lubrication (2) could still be used. However, to ensure  reliability, a precautionary measure is taken by opting for an oil bath (3).

**The solution to question c. is:**

- An oil bath is selected as the appropriate lubrication method for the chosen roller chain.

---

## d. Force on the drive shaft

The load on the drive shaft $F_A$ is slightly higher at the upper wheel due to the inclined position of the chain drive, corresponding to the difference between Fsn and Fso2. For the upper wheel, in accordance with equation 17.19, we have $F_A \approx F_t \cdot K_A + 2 \cdot F_{sb}$ , as per expression 17.17.  
The tangential force can be calculated using expression 17.14.


In [19]:
CH.F_sb = CH.E17_17_TopWheelSupportForce()
t=HM.MyHelp(CH.F_sb.evalf())

_g*_l_T*_q*(_Fprime_s + sin(0.005556*_psi*pi/deg_))

    q : [kg/m] chain weight / lenght
    psi : [deg] angle between horizontal and chain contact points line
    Fprime_s : [] relative load factor 
    l_T : [mm] length chain section under tension
    g : [m/s²] gravitationl constant, to be set by user


In [20]:
CH.q = CH.q = 2.7 * kg_/m_          # based on tabel 17-1
CH.Fprime_s = 5                     # based on tabel 17-4
CH.g = 9.81*m_/s_**2                # the chain will be used on Earth
CH.epsilon_0 = CH.E17_17_hB_epsilon_0()
t=HM.MyHelp(CH.epsilon_0.evalf())   #---epsilon_0 is already fully known --- 

asin(0.07874*(_d_2 - 186.5*mm_)/(mm_*(80.0*(1 - 0.36125/pi**2)**0.5 + 80.0)))

    d_2 : [] pitch diameter of the driven chain wheel


In [21]:
CH.d_2 = HM.substitute(CH.E17_3B_PitchDiameter(),CH,z=CH.z_2)       #---using the same expression, but for a different value for z ---
HM.EqPrint('d_2',CH.d_2)
CH.epsilon_0 = HM.substitute(CH.epsilon_0,CH)

Eq(d_2, 461.1*mm_)

*Can you observe how the document's overall structure follows a bottom-up approach, as it reflects the sequence of sub-questions posed?*

In [22]:
CH.F_t = CH.E17_14A_TangentialForce()   
CH.F_t = UM.All_to_SI(CH.F_t)
CH.F_t = UM.kgm_s2_to_N(CH.F_t)
t=HM.EqPrint('F_t',CH.F_t)

Eq(F_t, 2457.0*N_)

*Let's see which parameters are yet to be determined*

In [23]:
CH.F_sb = HM.substitute(CH.F_sb,CH)
t=HM.MyHelp(CH.F_sb.evalf())

26.49*_l_T*kg_*(sin(0.005556*_psi*pi/deg_) + 5.0)/s_**2

    psi : [deg] angle between horizontal and chain contact points line
    l_T : [mm] length chain section under tension


Equation 17.17 facilitates the determination of the support force acting on the upper sprocket, but it requires the configuration of several parameters before it can be applied

In [24]:
CH.l_T = CH.a * sp.cos(CH.epsilon_0)
# HM.EqPrint('l_T',CH.l_T.evalf())
CH.delta = 40*deg_       # as given in the assignemt
CH.psi = CH.E17_17_hA_psi()

CH.F_sb = HM.substitute(CH.F_sb,CH)
CH.F_sb = UM.All_to_SI(CH.F_sb)
CH.F_sb = UM.kgm_s2_to_N(CH.F_sb)
t=HM.EqPrint('F_sb',CH.F_sb)

Eq(F_sb, 146.1*N_)

Compared to the circumferential force $F_t$, the support tension $F_{sb}$ is relatively small and can be readily disregarded in this scenario.  
Taking into consideration the operating factor $K_A \approx 1.6$ (refer to part b), the shaft load on the upper sprocket is calculated as:

In [25]:
CH.F_Ab = CH.E17_19B_ShaftSupportForceTopChainWheel()
t=HM.EqPrint('F_Ab',CH.F_Ab)

Eq(F_Ab, 4224.0*N_)

**The solution to question d. is:**  
  
- The support force for the upper sprocket will be $F_{Ab} \approx 4225$ N  


---  

below you can find an example on how to incorporate images in a jupyternotebook file:

![image of chain](./images/ChainExample.jpg)